### ISSUE TITLE
Consider dropping `stdcall` usage in native host interfaces for .NET 7

### ISSUE THREAD

**Issue by alexrp** [dotnet/runtime#57586]
AFAICT, this platform difference is arbitrary and clearly adds friction. In some languages/ecosystems, it's not easy to achieve the goal of "export this function as `stdcall`, but actually, don't mangle the name as if it is `stdcall`" (see e.g. https://github.com/ziglang/zig/issues/9575).

I understand that there may be compatibility concerns with doing this, which is why I would suggest doing this in .NET 7.

Context:

* https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/tools/aot/jitinterface/dllexport.h#L15-L22
* https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/hosts/inc/coreclrhost.h#L11-L15
* https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/mono/mono/mini/main-core.c#L13-L25
* https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/native/corehost/nethost/nethost.h#L9-L37

**Comment by vitek-karas** [comment]
@AaronRobinsonMSFT you probably know exactly why and what could go wrong if we removed it, right?

**Comment by AaronRobinsonMSFT** [comment]
Sigh. Yes, this would be a breaking change for all win-x86 customers. Not saying we can't do it but we should have a serious conversation about our and user expectations. The reason for `stdcall` is historical and there is no real reason it has to be that at this point.

Here are the things we should think about.
- How impactful is an win-x86 breaking change really going to be? We would need to determine the user count for win-x86 and do a gut check on whether that number is big enough to care about this sort of breaking change.
- Breaking this will cause serious headaches for some people since we've changed the symbol and therefore it looks like `cdecl` but it is really `stdcall`. What this means in practice is someone could simply acquire the export via a call to [`GetProcAddress`](https://docs.microsoft.com/windows/win32/api/libloaderapi/nf-libloaderapi-getprocaddress), call it incorrectly, and thus corrupt the stack.
- The other side of the "small number of people using win-x86" argument is it isn't worth the potential fallout/effort/resourcing to changing this. Basically, regardless of our best-effort someone will be impacted.
- If there is a large number of win-x86 users I'd argue this isn't worth a change because `stdcall` is the Win32 API surface anyways so I'd like to understand why .NET is impacted but Win32 wouldn't be. I suppose it is possible people just want .NET and none of the Win32 API, but that doesn't change the fact that the platform we are running is that way and we are simply following the existing pattern.

I agree this is something to at least consider.

**Comment by jkotas** [comment]
> In some languages/ecosystems, it's not easy to achieve the goal of "export this function as `stdcall`, but actually, don't mangle the name as if it is `stdcall`" (see e.g. https://github.com/ziglang/zig/issues/9575).

Would this problem be addressed if we exported the same symbols both with and without stdcall mangling on x86? (It is only exported without stdcall mangling on x86 right now.)

**Comment by alexrp** [comment]
> Would this problem be addressed if we exported the same symbols both with and without stdcall mangling on x86? (It is only exported without stdcall mangling on x86 right now.)

If the host resolver would try to load the mangled name also, that would be sufficient for my case. But given this whole dance around `stdcall` mangling, I assumed there might be some very compelling reason why it doesn't.

**Comment by jkotas** [comment]
> If the host resolver would try to load the mangled name also, 

Why would need the host resolver try to load the mangled name also if we exported both mangled and unmangled name?

**Comment by alexrp** [comment]
I think I'm confused about what you're suggesting.

The issue in my case is that it's very difficult/annoying to export the unmangled name due to toolchain conventions (`stdcall` implies mangling). So either:

1. The toolchain I'm working with (Zig) needs to grow a way to override the calling convention mangling. It's not yet clear if that's even in the cards, let alone when it would happen.
2. The host resolver needs to try to load both mangled and unmangled names. In this case, I can export the mangled names normally. This has the benefit of also not breaking any existing code that's doing `stdcall` + unmangled names.
3. The host resolver and runtimes are changed to drop `stdcall` entirely and everyone agrees on `cdecl` everywhere. Might (?) break external users.

**Comment by jkotas** [comment]
> The toolchain I'm working with (Zig) needs to grow a way to override the calling convention mangling

Are you trying to build coreclr.dll using the Zig toolchain; or are you just trying to consume the standard coreclr.dll from a binary built by the Zig toolchain?

**Comment by alexrp** [comment]
> Are you trying to build coreclr.dll using the Zig toolchain; or are you just trying to consume the standard coreclr.dll from a binary built by the Zig toolchain?

The former, for a custom runtime experiment. (But consuming it would probably have the exact same issue, just from the other end.)

**Comment by vitek-karas** [comment]
Ever since .NET Core 3.0 we recommended using the `hostfxr` APIs to host .NET from native processes. Those changes simply remove samples/guides for the direct hosting of CoreCLR which has several problems. This has no impact on the compatibility  of the `hostfxr` APIs.

**Comment by alexrp** [comment]
Right, this isn't about `hostfxr` APIs, it's about removing `stdcall` from the `coreclr_*` APIs. I'm saying I think the compatibility concerns with doing so aren't as significant as I thought, since people shouldn't be using the `coreclr_*` APIs directly as of 3.0 anyway.

**Comment by tannergooding** [comment]
This sounds like a case where it would be fine to dual export.

`stdcall` does mangle by default (`float __stdcall AddFloats(float x, float y)` becomes `_AddFloats@8`). However, the Win32 SDK itself strips this mangling using a module-definition file (https://docs.microsoft.com/en-us/cpp/build/reference/module-definition-dot-def-files?view=msvc-160) and we could do the same while preserving `stdcall` and preserving back-compat.

That is, imagine you have:
```cpp
extern "C"
{
  float __stdcall AddFloats(float x, float y)
  {
    return x + y;
  }
}
```

By default in this scenario you get an export `_AddFloats@8` and so you get the following:
```
  auto lib = LoadLibrary(L"Dll1.dll");

  auto add1 = GetProcAddress(lib, "AddFloats");     // nullptr
  auto add2 = GetProcAddress(lib, "_AddFloats");    // nullptr
  auto add3 = GetProcAddress(lib, "_AddFloats@8");  // valid
  auto add4 = GetProcAddress(lib, "AddFloats@8");   // nullptr
```

You can however, define the following and pass it to the linker as `/def:Exports.def`:
```
LIBRARY   Dll1
EXPORTS
   _AddFloats@8
   AddFloats=_AddFloats@8
```

That causes (where `add1 == add3`):
```
  auto lib = LoadLibrary(L"Dll1.dll");

  auto add1 = GetProcAddress(lib, "AddFloats");     // valid
  auto add2 = GetProcAddress(lib, "_AddFloats");    // nullptr
  auto add3 = GetProcAddress(lib, "_AddFloats@8");  // valid
  auto add4 = GetProcAddress(lib, "AddFloats@8");   // nullptr
```

**Comment by vitek-karas** [comment]
My understanding is that this is not about mangling - coreclr exports are not mangled as far as I can tell. I think this issue is about calling convention, right?

For example, when `hostpolicy` loads `coreclr` it does `GetProcAddress` on unmangled names: https://github.com/dotnet/runtime/blob/8b94b58b894b286edfc36a02e6890bc6656b8b54/src/native/corehost/hostpolicy/standalone/coreclr_resolver.cpp#L21-L24

**Comment by AaronRobinsonMSFT** [comment]
>My understanding is that this is not about mangling - coreclr exports are not mangled as far as I can tell. I think this issue is about calling convention, right?

It is hard to separate those concerns actually. The `stdcall` calling convention "mangles" the symbol to indicate the calling contract (i.e., 'Callee will subtract @N bytes from the stack'). In this sense the mangle indicates the `stdcall` convention. Without that the symbol is assumed to be `cdecl` because of history. Basically mangling and calling convention in this case are coupled. We are currently exporting a `stdcall` function without the conventional symbol format so it is confusing to everyone outside of Microsoft.

**Comment by tannergooding** [comment]
> We are currently exporting a stdcall function without the conventional symbol format so it is confusing to everyone outside of Microsoft.

If we are already exporting it "unmangled", I don't see the issue. This is exactly what the entire Windows SDK does. `GetProcAddress` itself is `__stdcall` too and by default would be named `_GetProcAddress@8`, but its not. It's exported "unmanagled" as `GetProcAddress`.

**Comment by alexrp** [comment]
@tannergooding 

> This sounds like a case where it would be fine to dual export.

Just to be clear, the problem for me is from the *implementation* side, i.e. I don't have the ability to export a function that is `stdcall` while being unmangled. So I need the host resolver to change somehow (see https://github.com/dotnet/runtime/issues/57586#issuecomment-901276874). I don't particularly care whether we do (2) or (3), though in light of https://github.com/dotnet/runtime/issues/57586#issuecomment-909328096, I would argue that (3) is probably fine?

**Comment by AaronRobinsonMSFT** [comment]
>Just to be clear, the problem for me is from the implementation side, i.e. I don't have the ability to export a function that is stdcall while being unmangled.

Exactly. That is what I meant by "it is confusing to everyone outside of Microsoft". This is common for us but the rest of the world doesn't do that because it is confusing.

**Comment by tannergooding** [comment]
So the issue is that the tooling can "import" mangled or unmanagled stdcall. However it can only "export" managled stdcall?

Wouldn't the fix still be similar, if we export both mangled and unmangled, we can also have `pal::get_symbol` (or w/e we use) attempt to lookup both names as well? This would support custom hosts defined in other languages that can't declare unmangled stdcall.

**Comment by AaronRobinsonMSFT** [comment]
>Wouldn't the fix still be similar

Oh absolutely. Exporting both is a reasonable solution for sure. I was merely providing context for @vitek-karas that this is both a calling convention and mangling issue because of their tight coupling in this case.

**Comment by alexrp** [comment]
> So the issue is that the tooling can "import" mangled or unmanagled stdcall. However it can only "export" managled stdcall?

I think that's the case, yes. The Zig tooling doesn't seem to struggle with linking against e.g. Win32 APIs, but several people besides myself have run into the issue of exporting `stdcall` w/o mangling and there seems to be no solution as of yet.

> Wouldn't the fix still be similar, if we export both mangled and unmangled, we can also have `pal::get_symbol` (or w/e we use) attempt to lookup both names as well? This would support custom hosts defined in other languages that can't declare unmangled stdcall.

This would be option (2) I mentioned above. Though if we go down this route, I see no particular reason why CoreCLR and Mono should export both mangled and unmangled symbols - they would continue to just work as they are now. The change would just be in the host resolver.

**Comment by vitek-karas** [comment]
That actually sounds much less problematic (the host would obviously prefer the unmangled name first, but it could fall back to the mangled one). Sorry for not understanding the problem in your original description.

Above you mention that you're trying to do this to experiment with custom built runtime. If you're already rebuilding the entire coreclr, would it make sense to rebuild the host as well - in that case you can tweak it to your liking.
Alternatively you can just build a custom version of hostpolicy which looks up the mangled names (but still exports unmangled), so that you can use such runtime with other hosts. (hostpolicy is for all intents and purposes part of the runtime, so you would need to package it with your coreclr anyway).

**Comment by alexrp** [comment]
By custom runtime, I mean I'm implementing a runtime from scratch. When creating the platform packages for my runtime's SDK, I fish hostfxr/hostpolicy out of the official .NET packages and repackage them. (Not ideal, I know, but it seems to work fine so far, `stdcall` issue notwithstanding.)

**Comment by vitek-karas** [comment]
@alexrp - based on the above, he "agreed upon" solution is for the host to try to import both mangled and unmangled names from coreclr.dll. If you can, please send a PR with that change.

**Comment by alexrp** [comment]
Great, I'll try to find time to do a pull request ASAP.
